pkg/sql/sem/tree/constant.go

```diff
@@ -545,10 +550,29 @@ func (expr *StrVal) ResolveAsType(ctx *SemaContext, typ *types.T) (Datum, error)
 		return &expr.resString, nil
 	case types.BytesFamily:
 		return ParseDByte(expr.s)
-	}
 
-	datum, err := ParseAndRequireString(typ, expr.s, ctx)
-	return datum, err
+	case types.ArrayFamily,
```

Comment:
Is array included in here because it could be an array of these "bad" types? If so, can we make this check more granular to avoid doing the array cast at runtime for say int arrays?